### ISSUE TITLE
Type-matching arbitrary things against Byte/Short/Int ends up calling toString

### ISSUE THREAD

**Issue by lihaoyi** [scala-js/scala-js#2953]
```scala
object Walker{
  def treeify(x: Any): Unit = {
    println("Treeify!")
    x match{
      case x: Short => println("short")
      case x: Int => println("int")
      case x: Iterable[_] => println("iterable ")
    }
  }
}

class C(){
  var counter = 0
  override def toString = {
    counter += 1
    println("increment! " + counter)
    "C"
  }
}

Walker.treeify(List.fill(4)(new C))
// Treeify!
// increment! 1
// increment! 1
// increment! 1
// increment! 1
// increment! 2
// increment! 2
// increment! 2
// increment! 2
// iterable
```
https://scalafiddle.io/sf/OkgMiGx/0

It seems that the `x: Short =>` and `x: Int =>` cases are calling `.toString` on my Cs
`x: Byte =>` is doing it too (adding that makes the increments go up to 3) but Float, Double and Char don't seem to do it
on the JVM the increment messages don't appear at all

> Using arithmetic operators causes to `ToPrimitive(x, "number")` internal function of the JS spec to be called, which, in the absence of a valueOf method, will fall back on calling toString().
> We must assert that `typeof v === "number"` before trying any operator.
> - sjrd

**Comment by sjrd** [comment]
I'm amazed this was never discovered before ...

**Comment by lihaoyi** [comment]
I suppose most people's `.toStrings` are cheap/pure and so you wouldn't notice it being called redundantly. I only noticed because my PPrint library has tests to enforce laziness, making sure it doesn't `.toString` more-things/things-more-times than I expect to

**Comment by sjrd** [comment]
Fixed in eb0ca6e5ea930a6f543224c2aa41a71cffaee7df
